src/ppx_deriving_protobuf.cppo.ml

```diff
@@ -1170,15 +1170,20 @@ let rec write_protoc ~fmt ~path:base_path ?(import=[])
       let i = int_of_string sn in
 #endif
       Format.fprintf fmt " [default=%d]" i
-    | Some { pexp_desc = Pexp_constant (Pconst_string (s, _)) } ->
-      Format.fprintf fmt " [default=\"%s\"]" (escape ~pass_8bit:true s)
-    | Some [%expr Bytes.of_string [%e? { pexp_desc = Pexp_constant (Pconst_string (s, _)) }]] ->
-      Format.fprintf fmt " [default=\"%s\"]" (escape ~pass_8bit:false s)
     | Some { pexp_desc = Pexp_construct ({ txt = Lident n }, _) }
     | Some { pexp_desc = Pexp_variant (n, _) } ->
       Format.fprintf fmt " [default=%s_tag]" n
     | None -> ()
-    | Some { pexp_loc } -> raise (Error (Pberr_ocaml_expr pexp_loc))
+    | Some expr ->
+        let expr, pass_8bit =
+          match expr with
+          | [%expr Bytes.of_string [%e? sub_expr ]] -> sub_expr, false
+          | _ -> expr, true in
+        match Ppx_deriving.string_of_expression_opt expr with
```

Comment:
Naive question: we would *not* need `Ppx_deriving.string_of_expression` if there was support in metaquot on binding string literals. (We could also get rid of the ifdef above for integer constants, etc.). Is this available?

Reply:
This is not available: anti-quotations in patterns are limited to `[%e? ... ]` (expression), `[%t? ... ]` (type) and `[%p? ... ]` (pattern). It could be a nice addition: would you fill an issue? It looks quite boring to implement in `ppx_tools`, as AFAIK, there is one branch to change per supported OCaml version... Anyway, I think that `Ppx_deriving.string_of_expression_opt` would still be useful in some other contexts (when we are not in a pattern-matching, or not using metaquot anywhere else).

Reply:
There are several versions of ppx_tools around (ppx_tools, ppx_tools_versioned, and I wonder if ppxlib has its own reimplementation), so we would have to be careful if we wanted to implement this globally.

Reply:
ppxlib _has_ its own reimplementation (https://github.com/ocaml-ppx/ppxlib/blob/master/metaquot/ppxlib_metaquot.ml)... and I have mine (https://github.com/thierry-martinez/metaquot): I think it is worth proposing your extension at least for some of them (I am quite urged to implement it on mine, for instance!), but yes, the conclusion is that we cannot rely on that to replace `Ppx_deriving.string_of_expression_opt`.

Reply:
Thinking more about it, I am not sure what the interface should be. Ideally we would want `[%constr ...]`, but this is not explicit about the type of constant, so it is not very useful. I guess we could have `%lit.string`, `%lit.int`, `%lit.float`, `%lit.int32`? (For strings, do we provide provisions to match on the delimiters of quoted string literals?). There is a risk that we would end up reinventing a type of structured constants, encoded through the extension-name syntax.

Reply:
Yes, when I read your question, I thought about one extension-name per function in `Ast_helper.Const`. Even if it is a bit more cryptic, I would follow the convention of one-char names: `[%c ... ]`, `[%s ... ]`, `[%i ... ]`, `[%n ... ]`, `[%f ... ]`, with exceptions for `[%i32 ... ]` and `[%i64 ... ]`, and perhaps `[%integer ... ]` for matching an integer literal by its string representation. It is quite ad-hoc, indeed, but not much than the rest of metaquot...

Reply:
I think that cryptic is bad here: we don't want to pollute the extension namespace unnecessarily (maybe there will be a need for another quotation named `s` in metaquot later, which is used more often or tends to nest more than string literals), and I prefer to make the code reasonably easy to read.

Reply:
I'll open an issue in ppx_tools.

Reply:
(The issue is https://github.com/ocaml-ppx/ppx_tools/issues/83 )